File: younghyo/메뉴 리뉴얼.java
// [프로그래머스] 메뉴 리뉴얼
// 조합...

import java.util.*;

class Solution {
    static HashMap<String, Integer> hm;
    public String[] solution(String[] orders, int[] course) {
        hm = new HashMap<>();

        for(String o: orders) {
            char[] order = o.toCharArray();
            Arrays.sort(order);
            for (int j : course) {
                combination(0, 0, j, order, "");
            }
        }

        int[] maxCntOfCourse = new int[course[course.length -1] +1];
        for(String key: hm.keySet()) {

            if(hm.get(key) >= 2) {
                maxCntOfCourse[key.length()] = Math.max(hm.get(key), maxCntOfCourse[key.length()]);
            }
        }

        List<String> answerTemp = new ArrayList<>();

        for(String key: hm.keySet()) {
            for(int i=0; i< maxCntOfCourse.length; i++) {
                if(maxCntOfCourse[i] == 0) continue;
                if(key.length() == i && hm.get(key) == maxCntOfCourse[i]) {
                    answerTemp.add(key);
                }
            }
        }

        Collections.sort(answerTemp);
        String[] answer = new String[answerTemp.size()];
        int i=0;
        for(String str: answerTemp) {
            answer[i++] = str;
        }

        return answer;
    }

    private static void combination(int cnt, int start, int limit, char[] order, String result) {
        if(cnt == limit) {
            hm.put(result, hm.getOrDefault(result, 0)+1);
            return;
        }

        for(int i=start; i< order.length;i++) {
            combination(cnt+1, i+1, limit, order, result+order[i]);
        }
    }
}

File: jehyeok/메뉴리뉴얼.py
orders = ["XYZ", "XWY", "WXA"]
course = [2, 3, 4]

from itertools import combinations

def solution(orders, course):
    answer = []

    for c in course:
        maxCnt = 0
        result = []

        for order in orders:
            for data in combinations(order, c):
                tmp = list(data)
                cnt = 0
                for o in orders:
                    test = [x for x in o if x in tmp]

                    if len(test) == c: cnt += 1

                if cnt <= 1:
                    continue
                if cnt == maxCnt:
                    tmp.sort()
                    result.append(''.join(tmp))
                elif cnt > maxCnt:
                    tmp.sort()
                    result = [''.join(tmp)]
                    maxCnt = cnt

        for data in list(set(result)):
            answer.append(data)

    answer.sort()

    return answer

print(solution(orders, course))